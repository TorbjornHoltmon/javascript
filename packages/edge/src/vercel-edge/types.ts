import type { Session, User } from '@clerk/backend-core';
import type { GetSessionTokenOptions } from '@clerk/types';
import type { NextFetchEvent, NextRequest, NextResponse } from 'next/server';

export type WithAuthOptions = {
  loadUser?: boolean;
  loadSession?: boolean;
};

export type WithAuthMiddlewareCallback<Return, Options> = (
  req: RequestWithAuth<Options>,
  event: NextFetchEvent,
) => Return;

export type RequestWithAuth<Options extends WithAuthOptions = any> =
  NextRequest & {
    auth?: MiddlewareAuth;
  } & (Options extends { loadSession: true }
      ? { session: Session | null }
      : {}) &
    (Options extends { loadUser: true } ? { user: User | null } : {});

type NextMiddlewareResult = NextResponse | Response | null | undefined;

export type WithAuthNextMiddlewareHandler = (
  req: RequestWithAuth,
  event: NextFetchEvent,
) => Promise<NextMiddlewareResult>;

export type MiddlewareAuth = {
  sessionId: string | null;
  userId: string | null;
  getToken: (
    options?: GetSessionTokenOptions,
  ) => Promise<string | null> | string | null;
};
