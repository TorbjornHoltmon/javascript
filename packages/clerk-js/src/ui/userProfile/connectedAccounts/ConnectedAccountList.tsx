import { List } from '@clerk/shared/components/list';
import { TitledCard } from '@clerk/shared/components/titledCard';
import { OAuthProvider, OAuthStrategy } from "@clerk/types";
import React, { useState } from 'react';
import { Error } from 'ui/common/error';
import { useCoreUser, useEnvironment } from 'ui/contexts';
import { useNavigate } from "ui/hooks";
import { PageHeading } from 'ui/userProfile/pageHeading';

import { ConnectedAccountListItem } from './ConnectedAccountListItem';
import { UnconnectedAccountListItem } from './UnconnectedAccountListItem';

export function ConnectedAccountList(): JSX.Element {
  return <>
    <PageHeading title='Connected accounts' backTo='../' />

    <TitledCard className='cl-themed-card cl-list-card'>
      <ConnectedAccountListRows/>
    </TitledCard>
  </>;
}

function ConnectedAccountListRows(): JSX.Element {
  const [error, setError] = useState<string | undefined>();
  const user = useCoreUser();
  const { navigate } = useNavigate();
  const { userSettings: { social } } = useEnvironment();

  const availableProviders = Object.values(social).filter(oauthProvider => oauthProvider.enabled);

  const connectedProviders = user.externalAccounts.map(externalAccount => externalAccount.provider);

  const unconnectedProviders = availableProviders.filter((oauthProvider) => {
    const provider = oauthProvider.strategy.replace('oauth_', '') as OAuthProvider; // :-(
    return !connectedProviders.includes(provider);
  });

  const handleConnect = (strategy: OAuthStrategy) => {
    setError(undefined);

    user.connectExternalAccount({ strategy: strategy, redirect_url: window.location.href })
      .then(verification => {
        navigate(verification.externalVerificationRedirectURL);
      }).catch(err => {
        setError(err.message || err);
        console.log(err);
    });
  }

  if (availableProviders.length == 0) {
    return (
      <div className='cl-empty-list-item'>
        There are no available external account providers
      </div>
    );
  }

  return <>
    <Error>{error}</Error>

    <List>
      {user.externalAccounts.map(externalAccount => (
        <ConnectedAccountListItem
          key={externalAccount.id}
          externalAccount={externalAccount}
        />
      ))}

      {unconnectedProviders.map(unconnectedProvider => (
        <UnconnectedAccountListItem
          key={unconnectedProvider.strategy}
          oauthProviderSettings={unconnectedProvider}
          handleConnect={handleConnect}
        />
      ))}
    </List>
  </>
}
