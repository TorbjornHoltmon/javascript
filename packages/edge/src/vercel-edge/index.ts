import { AuthStatus, Base } from '@clerk/backend-core';
import { NextResponse } from 'next/server';

import { ClerkAPI } from './ClerkAPI';
import {
  WithAuthNextMiddlewareHandler,
  WithAuthMiddlewareCallback,
} from './types';
import { getAuthData } from './utils/getAuthData';

/**
 *
 * Required implementations for the runtime:
 * 1. Import Key
 * 2. Verify Signature
 * 3. Decode Base64
 * 4. ClerkAPI export with fetcher implementation
 * 5. Fetch Interstitial
 *
 */

const importKey = async (jwk: JsonWebKey, algorithm: Algorithm) => {
  return await crypto.subtle.importKey('jwk', jwk, algorithm, true, ['verify']);
};

const verifySignature = async (
  algorithm: Algorithm,
  key: CryptoKey,
  signature: Uint8Array,
  data: Uint8Array,
) => {
  return await crypto.subtle.verify(algorithm, key, signature, data);
};

const decodeBase64 = (base64: string) => atob(base64);

/** Base initialization */

const vercelEdgeBase = new Base(importKey, verifySignature, decodeBase64);

/** Export standalone verifySessionToken */

export const verifySessionToken = vercelEdgeBase.verifySessionToken;

/** Export ClerkBackendAPI API client */

export { ClerkAPI } from './ClerkAPI';

async function fetchInterstitial() {
  const response = await ClerkAPI.fetchInterstitial<Response>();
  return response.text();
}

/** Export middleware wrapper */

type WithAuthOptions = {
  authorizedParties?: string[];
  loadUser?: boolean;
  loadSession?: boolean;
};

export function withAuth(
  handler: WithAuthNextMiddlewareHandler<WithAuthOptions>,
  options: WithAuthOptions = {
    authorizedParties: [],
    loadSession: false,
    loadUser: false,
  },
): WithAuthNextMiddlewareHandler {
  return async function clerkAuth(req, event) {
    const { status, interstitial, sessionClaims } =
      await vercelEdgeBase.getAuthState({
        cookieToken: req.cookies['__session'],
        clientUat: req.cookies['__client_uat'],
        headerToken: req.headers.get('authorization'),
        origin: req.headers.get('origin'),
        host: req.headers.get('host') as string,
        userAgent: req.headers.get('user-agent'),
        forwardedPort: req.headers.get('x-forwarded-port'),
        forwardedHost: req.headers.get('x-forwarded-host'),
        referrer: req.headers.get('referrer'),
        authorizedParties: options.authorizedParties,
        fetchInterstitial,
      });

    if (status === AuthStatus.SignedOut) {
      return handler(req, event);
    }

    if (status === AuthStatus.Interstitial) {
      return new NextResponse(interstitial, {
        headers: { 'Content-Type': 'text/html' },
        status: 401,
      });
    }

    if (status === AuthStatus.SignedIn) {
      const { auth, user, session } = await getAuthData(req, {
        ...sessionClaims,
        options,
      });

      req.auth = auth;
      req.user = user;
      req.session = session;

      return await handler(req, event);
    }
  };
}
