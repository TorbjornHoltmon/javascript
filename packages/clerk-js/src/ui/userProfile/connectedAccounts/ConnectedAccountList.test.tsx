import { renderJSON } from '@clerk/shared/testUtils';
import {
  ExternalAccountResource,
  UserResource,
  UserSettingsJSON,
  UserSettingsResource
} from '@clerk/types';
import { OAuthProviders } from "@clerk/types/src";
import { ExternalAccount } from 'core/resources/ExternalAccount';
import { UserSettings } from 'core/resources/UserSettings';
import React from 'react';

import { ConnectedAccountList } from './ConnectedAccountList';

const mockNavigate = jest.fn();
jest.mock('ui/hooks', () => ({
  useNavigate: () => {
    return {
      navigate: mockNavigate,
    };
  },
}));

const mockUseEnvironment = jest.fn();

jest.mock('ui/contexts', () => {
  return {
    useCoreUser: (): Partial<UserResource> => {
      return {
        id: 'user_1nQu4nZrhHEeolMMRhg4yERFYJx',
        username: null,
        firstName: 'Peter',
        lastName: 'Smith',
        externalAccounts: [
          new ExternalAccount({
            id: 'fbac_yolo',
            provider: 'facebook',
            approvedScopes: 'email',
            emailAddress: 'peter@gmail.com',
            firstName: 'Peter',
            lastName: 'Smith',
            externalId: '10147951078263327',
          } as ExternalAccountResource),
          new ExternalAccount({
            id: 'gac_swag',
            provider: 'google',
            approvedScopes:
              'email https://www.googleapis.com/auth/userinfo.email https://www.googleapis.com/auth/userinfo.profile openid profile',
            emailAddress: 'peter@gmail.com',
            firstName: 'Peter',
            lastName: 'Smith',
            externalId: '112567347150540108741',
          } as ExternalAccountResource),
        ],
      };
    },
    useEnvironment: () => mockUseEnvironment(),
  };
});

const environmentContext = {
  userSettings: new UserSettings({
    social: {
      oauth_google: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_google',
      },
      oauth_facebook: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_facebook',
      },
      oauth_github: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_github',
      },
      oauth_microsoft: {
        enabled: true,
        required: false,
        authenticatable: true,
        strategy: 'oauth_microsoft',
      },
      oauth_bitbucket: {
        enabled: false,
        required: false,
        authenticatable: true,
        strategy: 'oauth_bitbucket',
      },
      oauth_discord: {
        enabled: false,
        required: false,
        authenticatable: true,
        strategy: 'oauth_bitbucket',
      }
    } as OAuthProviders,
  } as UserSettingsJSON) as UserSettingsResource,
};

const emptyEnvironmentContext = {
  userSettings: new UserSettings({
    social: {
      oauth_google: {
        enabled: false,
        required: false,
        authenticatable: true,
        strategy: 'oauth_google',
      },
      oauth_facebook: {
        enabled: false,
        required: false,
        authenticatable: true,
        strategy: 'oauth_facebook',
      },
    } as OAuthProviders,
  } as UserSettingsJSON) as UserSettingsResource,
};

jest.mock('ui/router/RouteContext', () => {
  return {
    useRouter: () => {
      return {
        resolve: () => {
          return {
            toURL: {
              href: 'http://www.ssddd.com',
            },
          };
        },
      };
    },
  };
});

describe('<ConnectedAccountList/>', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders a list of Connected Accounts', () => {
    mockUseEnvironment.mockImplementation(() => environmentContext);
    const tree = renderJSON(<ConnectedAccountList/>);
    expect(tree).toMatchSnapshot();
  });

  it('renders an empty list if there are no enabled providers', () => {
    mockUseEnvironment.mockImplementation(() => emptyEnvironmentContext);
    const tree = renderJSON(<ConnectedAccountList/>);
    expect(tree).toMatchSnapshot();
  });
});
