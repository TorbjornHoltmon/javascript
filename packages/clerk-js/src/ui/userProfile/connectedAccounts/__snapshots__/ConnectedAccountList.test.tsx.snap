// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ConnectedAccountList/> renders a list of Connected Accounts 1`] = `
Array [
  <div
    className="cl-page-heading"
  >
    <a
      aria-label="Back Button"
      className="cl-back-button cl-back-button"
      onClick={[Function]}
    >
      <svg
        data-filename="../shared/assets/icons/arrow-left.svg"
      />
    </a>
    <div
      className="cl-text-container"
    >
      <h2
        className="heading h2 cl-title"
      >
        Connected accounts
      </h2>
      <p
        className="cl-subtitle"
      />
    </div>
  </div>,
  <div
    className="card card cl-themed-card cl-list-card"
  >
    <div>
      <button
        className="cl-list-item elementContainer button hoverable"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="listItem line"
        >
          <div
            className="start"
          >
            <div>
              <img
                alt="Facebook Account"
                className="cl-left-icon-wrapper"
                src="https://images.clerk.dev/static/facebook.svg"
              />
              peter@gmail.com
            </div>
          </div>
          <div
            className="iconContainer"
          >
            <svg
              data-filename="../shared/assets/icons/arrow-right.svg"
            />
          </div>
        </div>
      </button>
      <button
        className="cl-list-item elementContainer button hoverable"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="listItem line"
        >
          <div
            className="start"
          >
            <div>
              <img
                alt="Google Account"
                className="cl-left-icon-wrapper"
                src="https://images.clerk.dev/static/google.svg"
              />
              peter@gmail.com
            </div>
          </div>
          <div
            className="iconContainer"
          >
            <svg
              data-filename="../shared/assets/icons/arrow-right.svg"
            />
          </div>
        </div>
      </button>
      <button
        className="cl-list-item elementContainer button hoverable"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="listItem line"
        >
          <div
            className="start"
          >
            <div>
              <img
                alt="oauth_github"
                className="cl-left-icon-wrapper"
                src="https://images.clerk.dev/static/github.svg"
              />
              Connect 
              Github
               account
            </div>
          </div>
          <div
            className="iconContainer"
          >
            <svg
              data-filename="../shared/assets/icons/arrow-right.svg"
            />
          </div>
        </div>
      </button>
      <button
        className="cl-list-item elementContainer button hoverable"
        disabled={false}
        onClick={[Function]}
      >
        <div
          className="listItem line"
        >
          <div
            className="start"
          >
            <div>
              <img
                alt="oauth_microsoft"
                className="cl-left-icon-wrapper"
                src="https://images.clerk.dev/static/microsoft.svg"
              />
              Connect 
              Microsoft
               account
            </div>
          </div>
          <div
            className="iconContainer"
          >
            <svg
              data-filename="../shared/assets/icons/arrow-right.svg"
            />
          </div>
        </div>
      </button>
    </div>
  </div>,
]
`;

exports[`<ConnectedAccountList/> renders an empty list if there are no enabled providers 1`] = `
Array [
  <div
    className="cl-page-heading"
  >
    <a
      aria-label="Back Button"
      className="cl-back-button cl-back-button"
      onClick={[Function]}
    >
      <svg
        data-filename="../shared/assets/icons/arrow-left.svg"
      />
    </a>
    <div
      className="cl-text-container"
    >
      <h2
        className="heading h2 cl-title"
      >
        Connected accounts
      </h2>
      <p
        className="cl-subtitle"
      />
    </div>
  </div>,
  <div
    className="card card cl-themed-card cl-list-card"
  >
    <div
      className="cl-empty-list-item"
    >
      There are no available external account providers
    </div>
  </div>,
]
`;
